### PR TITLE
Source ranges are better than locations

### PR DESCRIPTION
**Pull request by rictic**
* Also gives the responsibility of producing SourceRanges from Nodes to the Document

Commits are pretty logical, but the end point also looks pretty skimmable.

### CODE REVIEW

**Comment by rictic** [comment]
FYI I've got a followup PR that removes `node` on ScannedFeatures, ordering instead by the `sourceRange.start`. Seems simpler, and should be faster and reduce memory usage.

**Comment by FredKSchott** [comment]
+1 for adding short jsdoc-style descriptions to some of these functions like `applyLocationOffset` and `correctSourceRange`. With `correctSourceRange` specifically I was having trouble understanding what exactly "correction" meant in the context of a source range.

File: src/ast/inline-document-descriptor.ts [hunk 1]
```diff
@@ -57,33 +57,39 @@ export class InlineParsedDocument<N> implements ScannedFeature {
...
 }
 
-export function correctSourceLocation(
-    sourceLocation: SourceLocation,
-    locationOffset?: LocationOffset): SourceLocation|undefined {
-  if (!locationOffset || !sourceLocation) {
-    return sourceLocation;
+export function correctSourceRange(
+    sourceRange: SourceRange, locationOffset?: LocationOffset): SourceRange|
+    undefined {
```
Comment (FredKSchott) [inline src/ast/inline-document-descriptor.ts]: did we decide to change to `SourceRange?`?

File: src/ast/inline-document-descriptor.ts [hunk 2]
```diff
@@ -57,33 +57,39 @@ export class InlineParsedDocument<N> implements ScannedFeature {
...
   }
 }
 
-export function correctSourceLocation(
-    sourceLocation: SourceLocation,
-    locationOffset?: LocationOffset): SourceLocation|undefined {
-  if (!locationOffset || !sourceLocation) {
-    return sourceLocation;
+export function correctSourceRange(
+    sourceRange: SourceRange, locationOffset?: LocationOffset): SourceRange|
```
Comment (FredKSchott) [inline src/ast/inline-document-descriptor.ts]: if sourceRange can be undefined (you check for this below) this should be changed to `sourceRange?`?

File: src/elements-format.ts [hunk 3]
```diff
@@ -15,9 +15,9 @@
 /**
  * The base interface, holding properties common to many nodes.
  */
-export interface Node {
+export interface Feature {
```
Comment (FredKSchott) [inline src/elements-format.ts]: Why this change?

File: src/generate-elements.ts [hunk 4]
```diff
@@ -143,25 +142,21 @@ function serializeAttribute(
   return attribute;
 }
 
-function resolveSourceLocationPath(
+function resolveSourceRangePath(
     elementPath: string,
-    sourceLocation: SourceLocation|undefined): SourceLocation|undefined {
-  if (!sourceLocation) {
+    sourceRange: ResolvedSourceRange|undefined): (SourceRange | undefined) {
```
Comment (FredKSchott) [inline src/generate-elements.ts]: `ResolvedSourceRange?`, `SourceRange?`

File: src/generate-elements.ts [hunk 5]
```diff
@@ -143,25 +142,21 @@ function serializeAttribute(
...
+  if (!sourceRange) {
     return undefined;
```
Comment (FredKSchott) [inline src/generate-elements.ts]: nit: `return undefined` -> `return;`

File: src/elements-format.ts (hunk 3 above)
Comment (rictic) [inline src/elements-format.ts]: `Node` was too general, and Feature is how the rest of the code refers to this kind of thing

File: src/ast/inline-document-descriptor.ts [hunk 6]
```diff
@@ -57,33 +57,39 @@ export class InlineParsedDocument<N> implements ScannedFeature {
...
+    undefined {
+  if (!locationOffset || !sourceRange) {
+    return sourceRange;
   }
-  const result: SourceLocation = {
-    line: sourceLocation.line + locationOffset.line,
-    // The location offset column only matters for the first line.
-    column: sourceLocation.column +
-        (sourceLocation.line === 0 ? locationOffset.col : 0),
+  return <SourceRange>{
```
Comment (FredKSchott) [inline src/ast/inline-document-descriptor.ts]: I don't think you want to cast here, I'm not 100% sure but I think this might silence errors where the object doesn't actually conform to SourceRange

File: src/html/html-document.ts [hunk 7]
```diff
@@ -43,4 +45,34 @@ export class ParsedHtmlDocument extends ParsedDocument<ASTNode, HtmlVisitor> {
...
+  sourceRangeForNode(node: ASTNode): SourceRange {
+    if (!node || !node.__location) {
+      return;
+    }
+    const location = node.__location;
+    if (isElementLocationInfo(location)) {
+      return {
+        file: this.url,
+        start:
+            {line: location.startTag.line - 1, column: location.startTag.col},
```
Comment (FredKSchott) [inline src/html/html-document.ts]: why -1 on line but not column?

File: src/html/html-document.ts [hunk 8]
```diff
@@ -43,4 +45,34 @@ export class ParsedHtmlDocument extends ParsedDocument<ASTNode, HtmlVisitor> {
...
+        line: location.line - 1,
+        column: location.col + (location.endOffset - location.startOffset)
+      }
+    };
+  }
+}
+
+function isElementLocationInfo(location: parse5.LocationInfo|
+                               parse5.ElementLocationInfo):
+    location is parse5.ElementLocationInfo {
```
Comment (FredKSchott) [inline src/html/html-document.ts]: woahhhhh I've never seen this before. Groovy

File: src/test/ast/inline-document-descriptor_test.ts [hunk 9]
```diff
@@ -14,41 +14,40 @@
...
+            correctPosition({line: 1, column: 1}, {line: 1, col: 1}));
       });
 
   test('does not modify its input', function() {
-    const input: SourceLocation = {line: 5, column: 5};
+    const input: Position = {line: 5, column: 5};
     const offset: LocationOffset = {line: 5, col: 5};
-    const expected: SourceLocation = {line: 10, column: 5};
-    assert.deepEqual(expected, correctSourceLocation(input, offset));
+    const expected: Position = {line: 10, column: 5};
```
Comment (FredKSchott) [inline src/test/ast/inline-document-descriptor_test.ts]: wouldn't this be `column:10`?

File: src/ast/element-descriptor.ts [hunk 10]
```diff
@@ -38,24 +38,25 @@ export class ScannedElement implements ScannedFeature, Resolvable {
...
 
   jsdoc?: jsdoc.Annotation;
 
   applyLocationOffset(locationOffset?: LocationOffset) {
     if (!locationOffset) {
       return;
     }
-    this.sourceLocation =
-        correctSourceLocation(this.sourceLocation, locationOffset);
+    this.sourceRange = correctSourceRange(this.sourceRange, locationOffset);
```
Comment (justinfagnani) [inline src/ast/element-descriptor.ts]: what would it take to push this upstream to be correct when the object is created? The parser needs to know the loationOffset probably...

File: src/ast/source-range.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+export interface SourceRange {
+  /* The resolved path to the file. */
+  file: string;
```
Comment (justinfagnani) [inline src/ast/source-range.ts]: shouldn't the file be implicit? A Feature should belong to a document, so if a Feature it should be possible to get SourceRange's file from it's Feature. IOW, we need something like `.document` on Features.

File: src/elements-format.ts [hunk 12]
```diff
@@ -31,16 +31,26 @@ export interface Node {
   metadata?: any;
 }
 
-export interface SourceLocation {
-  /** Line number, zero indexed. */
-  line: number;
-  /** Column number, zero indexed. */
-  column: number;
+export interface SourceRange {
```
Comment (justinfagnani) [inline src/elements-format.ts]: why is this here and in source-range.ts?

File: src/elements-format.ts [hunk 13]
```diff
@@ -31,16 +31,26 @@ export interface Node {
...
   /**
-   * Path to file, relative to the package base. If not present, is the
-   * element's file.
+   * Path to the file containing the definition of the feature,
+   * relative to the parent feature, or the package if the feature has no parent
+   * (e.g. elements).
+   *
+   * If blank, the feature is defined in the same file as its parent.
```
Comment (justinfagnani) [inline src/elements-format.ts]: When is a Feature not in the same file as its parent?

File: src/generate-elements.ts (hunk 4 above)
Comment (justinfagnani) [inline src/generate-elements.ts]: I think you mean `sourceRange?: ResolvedSourceRange`?

File: src/html/html-document.ts (hunk 7 above)
Comment (justinfagnani) [inline src/html/html-document.ts]: 0 vs 1 based lines? Add a comment :)

File: src/polymer/feature-finder.ts [hunk 14]
```diff
@@ -43,14 +45,15 @@ export function featureFinder() {
       return;
     }
 
-    for (const prop of featureNode.properties.map(
-             esutil.toScannedPolymerProperty)) {
+    const polymerProps = featureNode.properties.map(
+        p =>
```
Comment (justinfagnani) [inline src/polymer/feature-finder.ts]: style nit: I prefer always using parens on arrow functions (it's also a "tip" int he style guide)

File: src/ast/inline-document-descriptor.ts (hunk 1 above)
Comment (rictic) [inline src/ast/inline-document-descriptor.ts]: That only works on identifier names, not on type names, so AIUI this is the only syntax for annotating return values.

File: src/ast/inline-document-descriptor.ts (hunk 2 above)
Comment (rictic) [inline src/ast/inline-document-descriptor.ts]: Good catch, done

File: src/generate-elements.ts (hunk 4 above)
Comment (rictic) [inline src/generate-elements.ts]: Done

File: src/generate-elements.ts (hunk 5 above)
Comment (rictic) [inline src/generate-elements.ts]: Done, here and elsewhere.

File: src/ast/inline-document-descriptor.ts (hunk 6 above)
Comment (rictic) [inline src/ast/inline-document-descriptor.ts]: Good catch, done.

File: src/html/html-document.ts (hunk 7 above)
Comment (rictic) [inline src/html/html-document.ts]: Added a comment. I think this gets tweaked in a downstream commit.

File: src/html/html-document.ts (hunk 8 above)
Comment (rictic) [inline src/html/html-document.ts]: Yeah, super useful!

File: src/test/ast/inline-document-descriptor_test.ts (hunk 9 above)
Comment (rictic) [inline src/test/ast/inline-document-descriptor_test.ts]: It's initially counterintuitive, but no! We only should correct the column offset on the *first line* of the inline document. Otherwise the column offsets are correct.

File: src/test/ast/inline-document-descriptor_test.ts [hunk 15]
```diff
@@ -14,41 +14,40 @@
...
+    assert.deepEqual(expected, correctPosition(input, offset));
```
Comment (rictic) [inline src/test/ast/inline-document-descriptor_test.ts]: Ack, I always get that wrong. Fixed here and elsewhere in this file.

File: src/ast/element-descriptor.ts (hunk 10 above)
Comment (rictic) [inline src/ast/element-descriptor.ts]: +1. Added a TODO

File: src/ast/source-range.ts (hunk 11 above)
Comment (rictic) [inline src/ast/source-range.ts]: Aren't we operating on a principle of avoiding pointers back up? I'd also add that all current uses of source ranges want string file urls, which suggests to me that this is the more useful format.

File: src/elements-format.ts (hunk 12 above)
Comment (rictic) [inline src/elements-format.ts]: Here the file is optional, in source-range.ts it is not.

File: src/elements-format.ts (hunk 13 above)
Comment (rictic) [inline src/elements-format.ts]: Inheritance mostly

File: src/polymer/feature-finder.ts (hunk 14 above)
Comment (rictic) [inline src/polymer/feature-finder.ts]: Done. Can we get tslint to check for that?

**Comment by rictic** [comment]
All comments addressed, PTAL

**Comment by FredKSchott** [comment]
LGTM! 💥🚀🎉! 
(but wait for a LGTM from @justinfagnani before merging :)

File: src/elements-format.ts (hunk 13 above)
Comment (justinfagnani) [inline src/elements-format.ts]: I don't quite grok that. Is a "parent" the same as "super"? Wouldn't a parent be an enclosing Feature, like a Element in a Document?

Comment (rictic) [inline src/elements-format.ts]: This is a json format, so parent is the containing object. There's no `super` concept. 

This is what I mean:

```javascript
{
"elements": [{
  "tagName": "paper-button",
  "url": "paper-button.html",
  "properties": {
    "buttonState": {
      "raised": {
        /* no file, defined in paper-button.html */
        "start": {/*...*/},
        "end": {/*...*/}
      },
    "buttonState": {
      "sourceRange": {
        "file": "./bower_components/iron-button-behavior/iron-button-behavior.html",
        "start": {/*...*/},
        "end": {/*...*/}
      },
      /*...*/
    },
    /*...*/
  }

Comment (justinfagnani) [inline src/elements-format.ts]: I see.

File: src/ast/source-range.ts (hunk 11 above)
Comment (justinfagnani) [inline src/ast/source-range.ts]: I think the way we're doing resolution could mean that we don't need that rule for resolved objects, just for the things being cached. But SourceRange applies to more than resolved things, so...

**Comment by justinfagnani** [comment]
LGTM++

**Comment by rictic** [comment]
Manually squashed and merged onto master.
